src/PluginLoader.php

```
@@ -44,7 +40,12 @@ public function __construct(string $installedJson = null, array $plugins = null)
             @trigger_error('Passing the path to the Composer installed.json as first argument is no longer supported in version 2.3.', E_USER_DEPRECATED);
         }
 
-        $this->plugins = $plugins ?: [];
+        $this->plugins = null !== $plugins ? $plugins : array_filter((array) @include self::getGeneratedPath());
```

Using `@include` is an anti-pattern, because it will also suppress error messages inside the included files.

---

These are generated files. We're talking performance here. Essentially it's the same as https://github.com/symfony/dependency-injection/blob/7cfcb0e2747aa4ce7ba92cde8f4d98b8dfce00a0/Dumper/PhpDumper.php#L588.

---

No, it isn‘t? The `include_once` call in this line is not silenced.

---

So how am I supposed to include without an additional file check?

---

You are not. A file check will be necessary here.

---

IIRC using the error suppression variant is actually slower than a `file_exists` check in most cases. But I've got no source for that at hand.

---

Did a naive test out of interest. This is the time in µs for 1M iterations each (lower numbers are better):

```
    existing / @:      6.1252429485321
non-existing / @:      8.4501919746399
    existing / check:  8.349004983902
non-existing / check:  1.2823150157928
```

So seems it's marginally better to use `@` when the file exists but much faster with the check if it doesn't. But all in all the difference is extremely small and only shows when executed billions of times in a row.

---

Updated 👍